clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -1149,7 +1149,8 @@ public void flush() {
      * block forever.</strong>
      * <p>
      *
-     * @throws InterruptException If the thread is interrupted while blocked
+     * @throws InterruptException If the thread is interrupted while blocked.
+     * @throws KafkaException If a Kafka related error occurs that does not belong to the public API exceptions.
```

Comment:
Thanks for adding this documentation! Unfortunately, this text is a bit confounding.

As I read it, it says that the broker may return some errors, which are not documented as part of the public API. Therefore, the client just throws a generic exception, with no information about what happened, what it means, or what I can do about it. Is that good library design?

If I find myself with such an exception, I would be totally at a loss as to what to do next... I guess go read the broker logs and source code to find out what happened? If that's really the answer, then a generic exception is indeed appropriate. Also, this situation must be exceedingly rare and requiring human intervention on the brokers, right? Something like the broker running out of memory or disk?

Concretely, even if this is the gloomy situation right now, it's better to have it documented than not, but we should include _some_ guidance. Such as saying that these exceptions are fatal and unrecoverable, and that they indicate the broker is unhealthy and the client is in an unknown state. Or (if applicable), that the exception may or may not be fatal, and that the user could optimistically attempt to retry if they want to. Or something. I'm not asking you to exhaustively document the situation, just apply the knowledge you already have to briefly provide a little more information to users.

Future generations can take on the work of actually categorizing and documenting the exceptions that can get thrown.

Reply:
+1, I'm not sure how we could hit this exception in real life, but if we anticipate it could happen, there needs to plan action items to triage further.

Reply:
Sounds great, I will try to update it.